商品購入機能

k-foo opened the pull request:
# What
- Formオブジェクトを利用し、２つの各テーブルへデータを保存。
- 上記のモデル単体コードを作成。
- PAYJPを用いたクレジットカード決済機能の実装。
- sold outの表示。
- 出品者以外が購入できるかつ未購入の商品であれば購入できるように実装。
# Why
- 購入記録と発送先を分けることによってコントローラーとモデルの肥大化を防ぐため。
- バリデーションをした上で保存されるか確認するため。今後のメンテナンスのため。
-  クレジットカード決済機能はユーザーの利便性のため。PAYJPの使用はセキュリティ・手間削減のため。
- 購入済みであるか一目でわかるようにするため。
- 出品者が購入できたり、購入後の商品も購入選択できたりすることを防ぐため。

## 確認画像
<img width="1254" alt="【ER図】furima-29074" src="https://user-images.githubusercontent.com/68714247/102007762-2799ff00-3d6f-11eb-9e11-58e7b2c9d1d9.png">

  
## 確認動画
●フォームオブジェクトのモデル単体テストコードの実行
https://gyazo.com/1021b19f776e422227c9bc44520f454c

●ログインしていないユーザーは購入ページに遷移しようとすると、ログインページに遷移すること
https://gyazo.com/0447f4fb89c345f5991bdf61cfe168a7

●出品者はURLを直接入力して購入ページに遷移しようとすると、トップページに遷移すること
前半　https://gyazo.com/931932cca497dd1c8d724e6ab8d2739b
後半　https://gyazo.com/8e3482fb8037771f574aaa65a8f1a8c3

●URLを直接入力して購入済み商品の購入ページへ遷移しようとすると、トップページに遷移すること
https://gyazo.com/39be305d71be7db0471542e0a27d079d

●クレジットカード決済ができること
●購入が完了したら、トップページまたは購入完了ページに遷移すること
前半　https://gyazo.com/998c47ec6a91acc2aeed6be23a800906
後半　https://gyazo.com/998c47ec6a91acc2aeed6be23a800906

●クレジットカード情報は必須であり、正しいクレジットカードの情報で無いときは決済できないこと
https://gyazo.com/7119ac1f7fa6305812ae65aa8430099b

●購入した商品は、再度購入できない状態になっていること
https://gyazo.com/85def7a11367d9db88834821b6ab3554

●エラーハンドリングができていること（適切では無い値が入力された場合、情報は保存されず、エラーメッセージを出力させる）
https://gyazo.com/9f9545851daec4cc8895bc005936763c

k-foo committed:
READMEとER図の修正

モデルの名前をわかりやすい名詞にするため。

k-foo committed:
Formオブジェクト（order_form）の作成

k-foo committed:
購入機能のルーティング設定

商品があっての購入となるため入れ子構造にする。

k-foo committed:
都道府県は「prefecture_id」というカラム名に統一。

paymentsテーブル作成時に、itemsテーブルにある「delivery_area_id」（都道府県のActiveHash）の内容が一緒であることに気づいたため。全て「prefecture_id」に統一した。

k-foo committed:
orderモデルの作成

k-foo committed:
paymentモデルの作成

k-foo committed:
２つのテーブル作成に伴いuserテーブルのアソシエーション修正。

k-foo committed:
ordersテーブルのマイグレーション作成

k-foo committed:
paymentsテーブルのマイグレーション作成

k-foo committed:
ordersコントローラーの作成。

k-foo committed:
ordersビューの修正。

ダミーでは無く選択した商品を表示したり、form_withにorder_form（Formオブジェクト）を指定したり、エラーメッセージが表示されるようにしたり、name属性の修正を行ったりした。その後、ルボコップを実行。

k-foo committed:
ER図・READMEの修正

カラム名に修正があったため。

k-foo committed:
Rspecファイルの作成。

モデルの単体テストコードを作るため。また「prefecture_id」に統一した後、テストコードを実行し、不備がないか確認するため。

k-foo committed:
Formオブジェクト（order_form）のモデル単体テストの作成。

k-foo committed:
ルボコップの実行。

k-foo committed:
クレジット決済の機能実装。

k-foo committed:
sold outの表示・対象者のみが購入出来るよう実装。

k-foo committed:
修正漏れがあったため修正。

①カラム変更、②複数形誤り

k-foo committed:
READMEの修正。

ER図が上手く載せられていなかったため。

k-foo committed:
モデルの単体テストコードにtokenを追加。

k-foo committed:
ビューの条件分岐を修正。

「ログイン中かつ出品者かつ未購入品」の場合には商品編集・削除ができ、「ログイン中かつ自分以外の出品者かつ未購入品」の場合には購入できるよう実装。

k-foo committed:
sold outの商品はURLを直接入力しても購入できないよう実装。

i-date commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,42 @@
+
+  def create
+    @order_form = OrderForm.new(order_params)
+    @item.price = Item.find(params[:item_id]).price
```
11行目ですが、` @item`は`Item.find(params[:item_id])`を代入しているので、再度`@item.price`に`Item.find(params[:item_id]).price`を代入する必要はないと考えられます。

そのため、11行目の`@item.price = Item.find(params[:item_id]).price`を一度削除いただき、動作をご確認いただければと存じます。

k-foo replied:
ありがとうございます。削除しました！（動作確認しました）

i-date commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,106 @@
+  end
+
+  describe '配送先情報の保存' do
+    context '配送先情報の保存ができるとき' do
```
「配送先情報の保存ができるとき」のテストに「建物名は空でも購入できる」というテストを追加しましょう。
建物名は任意入力項目のため、空でも購入できることを確認する必要があります。

k-foo replied:
修正しました！

i-date commented on spec/models/order_form_spec.rb:
```diff
@@ -0,0 +1,106 @@
+        expect(@order_form.errors.full_messages).to include("Phone number can't be blank")
+      end
+      it '電話番号にハイフンがあると保存できないこと' do
+        @order_form.phone_number = 123 - 1234 - 1234
```
90行目の`123 - 1234 - 1234`は`"`で囲んでおきましょう。
現在は、ハイフン（`-`）ではなくて、引き算のマイナス（`-`）と読まれているかと存じます。

k-foo replied:
修正しました！

i-date reviewed (commented):
レビュー依頼ありがとうございます！
コメントいたしましたので、ご確認をお願いいたします。

erika618 commented:
確認いただきありがとうございました。

●修正後の、テストコードの実行
https://gyazo.com/cfdffd9370eec5e5c3538bf3aedc899f

●修正後の動作確認（`@item.price = Item.find(params[:item_id]).price`削除後の購入）
前半　https://gyazo.com/3f2d463b8dbec73b7c86e8e13e8d506e
後半　https://gyazo.com/da23d527f745a94f64a70dfd56f127b1

k-foo committed:
orderコントローラーにおける蛇足の削除。

`@item.price = Item.find(params[:item_id]).price`とは別に、`@item = Item.find(params[:item_id])`と定義済みのため。

k-foo committed:
テストコードの修正。

①任意である建物名が空でも保存できるか②電話番号は文字列のため（string型）’’で囲み、ハイフンがあると保存できないことを証明できるか。

k-foo committed:
soldoutの商品は編集・削除できないように実装。

TakumaKawashima reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう。

k-foo committed:
商品の編集・削除について、URLを直接入力しても購入後は編集できないよう実装。

editは編集できてしまっていたため。orderが抜けていたことにより表示がおかしくなってしまっていたため。

erika618 commented:
申し訳ありません。マージしようとしたらビューがおかしいことに気づき、
条件分岐間違えていたことに気がつきました。

お手数ですが、もう一度ご確認いただけると嬉しいです。
（特に、`app/views/items/show.html.erb`と、`app/controllers/items_controller.rb`、`app/controllers/orders_controller.rb`）

●（出品者が）sold out の商品をURLダイレクトに入力しても、ちゃんとルートパスに戻る
前半　https://gyazo.com/701407f1b268bf25e6741269f94bc0d3
後半　https://gyazo.com/450b12983d6c752752ed0552b7cc766a

●（出品者以外が）sold out の商品をURLダイレクトに入力しても、ちゃんとルートパスに戻る
前半　https://gyazo.com/757d03fb12ec1f2676f0113a5f5c64b7
後半　https://gyazo.com/757d03fb12ec1f2676f0113a5f5c64b7

何度もお手数かけて申し訳ありません！
どうぞよろしくお願いします。

ky3090 reviewed (commented):
再度レビュー依頼ありがとうございます。
自身で気づかれていて素晴らしいです！
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう。

yutakatahira commented:
> 再度レビュー依頼ありがとうございます。
> 自身で気づかれていて素晴らしいです！
> LGTMとさせていただきます。実装お疲れ様でした。
> 
> マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
> その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
> 引き続き実装頑張っていきましょう。

大変申し訳ございません。こちらに関しましてご放念くださいませ。

ky3090 commented on app/views/items/show.html.erb:
```diff
@@ -24,17 +25,14 @@
+     <%= link_to "売り切れました",item_path(@item),class:"disabled-button bold" %>
+    <% else  %>
+      <%= link_to '購入画面に進む', item_orders_path(@item), class:"item-red-btn"%>
   <% end %>
```
```
- ログアウト状態のユーザーには、「編集・削除・購入画面に進むボタン」が表示されないこと
```

現在の記述ですと、ログアウト状態で遷移すると購入画面に進むボタンが表示されてしまいますので
上記の要件を満たす記述に修正しましょう！

ky3090 reviewed (commented):
大変お手数おかけし、申し訳ございません。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

k-foo committed:
ログアウト中に購入画面URLをダイレクトに押しても購入できないよう実装。

erika618 commented:
見てくださってありがとうございます！（助かりました）
修正したのでお願いします。

●ログアウト中の場合に、購入画面に遷移できないよう実装。
https://gyazo.com/9886191aba4269e80d9a0f1e8654596f

k-foo committed:
ログアウト中の場合は、商品の編集・削除・購入が表示されないよう実装。

ログアウトしている方に対しては、商品の編集・削除・購入の表示を予め表示させない要件であったため。（修正前は、表示はされるが、編集・削除・購入はできない状態、という表示であった。）

TakumaKawashima reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認させていただきました。
LGTMとさせていただきます。実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。

コメントも残しながら、丁寧に学習されていますね。
引き続き実装頑張っていきましょう。
